dll/win32/kernel32/kernel32_vista/firmware.c

```diff
@@ -0,0 +1,231 @@
+                                               pdwAttribubutes);
+
+    /* Cleanup and return */
+    RtlFreeUnicodeString(&Namespace);
+    RtlFreeUnicodeString(&VariableName);
+    return Length;
+}
+
+BOOL
+WINAPI
+SetFirmwareEnvironmentVariableExW(
+    _In_ LPCWSTR lpName,
+    _In_ LPCWSTR lpGuid,
+    _In_reads_bytes_opt_(nSize) PVOID pValue,
+    _In_ DWORD nSize,
+    _In_ DWORD dwAttributes)
+{
+    NTSTATUS Status;
+    UNICODE_STRING VariableName, Namespace;
+    GUID VendorGuid;
+
+    /* Check input parameters and build NT strings */
+    if (lpName == NULL || lpGuid == NULL)
+    {
+        SetLastError(ERROR_INVALID_PARAMETER);
+        return FALSE;
+    }
+    RtlInitUnicodeString(&VariableName, lpName);
+    RtlInitUnicodeString(&Namespace, lpGuid);
+    Status = RtlGUIDFromString(&Namespace, &VendorGuid);
+    if (!NT_SUCCESS(Status))
+    {
+        BaseSetLastNTError(Status);
+        return FALSE;
+    }
+
+    /* Set firmware system environment variable value */
+    Status = NtSetSystemEnvironmentValueEx(&VariableName,
+                                           &VendorGuid,
+                                           pValue,
+                                           nSize,
+                                           dwAttributes);
+    if (!NT_SUCCESS(Status))
+    {
+        BaseSetLastNTError(Status);
+        return FALSE;
+    }
+
+    return TRUE;
+}
+
+BOOL
+WINAPI
+SetFirmwareEnvironmentVariableExA(
+    _In_ LPCSTR lpName,
+    _In_ LPCSTR lpGuid,
+    _In_reads_bytes_opt_(nSize) PVOID pValue,
+    _In_ DWORD nSize,
+    _In_ DWORD dwAttributes)
+{
+    NTSTATUS Status;
+    BOOL Result;
+    UNICODE_STRING VariableName, Namespace;
+    ANSI_STRING AnsiVariableName, AnsiNamespace;
+
+    /* Check input parameters and build NT strings */
+    if (lpName == NULL || lpGuid == NULL)
+    {
+        SetLastError(ERROR_INVALID_PARAMETER);
+        return FALSE;
+    }
+    RtlInitString(&AnsiVariableName, lpName);
+    Status = RtlAnsiStringToUnicodeString(&VariableName, &AnsiVariableName, TRUE);
+    if (!NT_SUCCESS(Status))
+    {
+        BaseSetLastNTError(Status);
+        return FALSE;
+    }
+    RtlInitString(&AnsiNamespace, lpGuid);
+    Status = RtlAnsiStringToUnicodeString(&Namespace, &AnsiNamespace, TRUE);
+    if (!NT_SUCCESS(Status))
+    {
+        RtlFreeUnicodeString(&VariableName);
+        BaseSetLastNTError(Status);
+        return FALSE;
+    }
+
+    /* Call unicode version interface */
+    Result = SetFirmwareEnvironmentVariableExW(VariableName.Buffer,
+                                               Namespace.Buffer,
+                                               pValue,
+                                               nSize,
+                                               dwAttributes);
+
+    /* Cleanup and return */
+    RtlFreeUnicodeString(&Namespace);
+    RtlFreeUnicodeString(&VariableName);
+    return Result;
+}
+
+_Success_(return)
```

Comment:
This annotation shouldn't be necessary (since return type is BOOL), should it?

Reply:
It should, because `BOOL` is `int` without SAL: `typedef int BOOL;`.

For example, MSVC will report warning if no such SAL:
![image](https://github.com/reactos/reactos/assets/55521781/26327edd-d0c8-45c6-83d4-3eee981afcf1)

And no warning if add this SAL:
![image](https://github.com/reactos/reactos/assets/55521781/675a33b1-356b-440d-b9f9-e8a9cc902294)